### PR TITLE
fix(cats-redis): release Semaphore to avoid leaking

### PR DESCRIPTION
**Pull request by Young-Zen**
### Preconditions to reproduce
The settings in [clouddriver.yaml](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/clouddriver-web/config/clouddriver.yml#L11) do apply: Set AgentScheduler type to ClusteredSortAgentScheduler and redis.parallelism greater than 0.
```
redis:
  scheduler: sort
  parallelism: 10 # greater than 0
```

### Additional context
In this [Code](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L416), the runningAgents has no release a permit to the Semaphore:
1. When worker is null or score is null;
2. When acquireAgent method throw an Exception.

### CODE REVIEW

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java [hunk 1]
```diff
@@ -409,17 +409,24 @@ private void saturatePool() {
...
+            // If on release of this agent, the scores don't match, this agent was rescheduled by a
+            // separate thread.
+            worker.setScore(score.acquireScore);
+            workers.add(worker);
+          } else {
+            runningAgents.ifPresent(Semaphore::release);
+          }
+        } catch (Throwable t) {
+          log.error("Failed to add AgentWorker", t);
+          runningAgents.ifPresent(Semaphore::release);
```
Comment (dbyron-sf) [inline cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:429]: Bettter to re-throw t to retain the current behavior?  I don't know enough about this code to know which way is better.

File: cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.netflix.spinnaker.cats.agent.ExecutionInstrumentation
+import com.netflix.spinnaker.cats.cluster.DefaultAgentIntervalProvider
+import com.netflix.spinnaker.cats.cluster.NodeStatusProvider
+import com.netflix.spinnaker.cats.test.TestAgent
+import redis.clients.jedis.Jedis
+import redis.clients.jedis.JedisPool
+import spock.lang.Specification
+import spock.lang.Subject
+
+class ClusteredSortAgentSchedulerSpec extends Specification {
```
Comment (dbyron-sf) [inline cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy:30]: It's perhaps too late for this PR, but we'd much prefer new code -- even test code -- in any language but groovy :)  Java or kotlin please.

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java (hunk 1 above)
Comment (Young-Zen) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:429]: Re-throw the Throwable will also cause runningAgents to not release permission to the Semaphore.
image

File: cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy (hunk 2 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy:30]: Ok, I will try.

Comment (Young-Zen) [reply cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy]: Done.

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java (hunk 1 above)
Comment (dbyron-sf) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:429]: I figured the additional throw would come after the release you added here, but maybe that still doesn't work.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  @Before
+  public void setUp() throws IllegalAccessException {
+    when(jedisPool.getResource()).thenReturn(jedis);
+    when(jedis.scriptLoad(anyString())).thenReturn("testScriptSha");
+    when(jedis.time()).thenReturn(List.of("1678784468", "374338"));
+    DefaultAgentIntervalProvider intervalProvider = new DefaultAgentIntervalProvider(6000000);
+    clusteredSortAgentScheduler =
+        new ClusteredSortAgentScheduler(jedisPool, () -> false, intervalProvider, parallelism);
+
+    WAITING_SET =
```
Comment (dbyron-sf) [inline cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:74]: Easier to make these fields package private in ClusteredSortAgentScheduler?  Marking them as `@VisibleForTesting` makes it pretty clear why...then we don't need a copy here.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "WAITING_SET", true)
+                .get(clusteredSortAgentScheduler);
+    WORKING_SET =
+        (String)
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "WORKING_SET", true)
+                .get(clusteredSortAgentScheduler);
+    saturatePoolMethod =
+        MethodUtils.getMatchingMethod(ClusteredSortAgentScheduler.class, "saturatePool");
+    saturatePoolMethod.setAccessible(true);
+    runningAgents =
```
Comment (dbyron-sf) [inline cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:85]: I'd say it's worth adding a getter for testing, but it's also nice to leave it hidden.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+              latch.countDown();
+              return null;
+            })
+        .when(agentExecution)
+        .storeAgentResult(any(), any());
+
+    saturatePoolMethod.invoke(clusteredSortAgentScheduler);
+    latch.await(10, TimeUnit.SECONDS);
+
+    assertEquals(runningAgents.map(Semaphore::availablePermits).orElse(-1), parallelism);
```
Comment (dbyron-sf) [inline cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:135]: I could use a comment here, specifically about the `orElse`.  Under what conditions does the orElse actually happen?

File: cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy (hunk 2 above)
Comment (dbyron-sf) [reply cats/cats-redis/src/test/groovy/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerSpec.groovy]: Thank you!

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java (hunk 1 above)
Comment (Young-Zen) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:429]: Yes, you're right. Before this PR, it released the acquired permit to the Semaphore [here](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L471), which will run in the worker thread managed by [agentWorkPool](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L427). In this way, before the agentWorker is submitted to the agentWorkPool, the acquired permit will not be released due to an exception throw. Regarding the question of whether to re-throw, I think it is better to ignore, so that the agentWorker can be submitted to the agentWorkPool as much as possible.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java (hunk 3 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:74]: Ok, I get it, thanks!

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java (hunk 5 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:135]: In this [code](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L109), when parallelism is -1, the runningAgents is Optional.empty().

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java (hunk 4 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:85]: Yeah, i prefer to leave it hidden.

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java [hunk 6]
```diff
@@ -59,7 +60,7 @@
  */
 public class ClusteredSortAgentScheduler extends CatsModuleAware
     implements AgentScheduler<ClusteredSortAgentLock>, Runnable {
-  private static enum Status {
+  private enum Status {
```
Comment (dbyron-sf) [inline cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:63]: Is this change necessary in this PR?

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  @Test
+  public void testRunningAgentsSemaphore() {
+    when(jedis.zrangeByScore(eq(ClusteredSortAgentScheduler.WORKING_SET), anyString(), anyString()))
+        .thenReturn(new HashSet<>());
+    when(jedis.zrangeByScore(eq(ClusteredSortAgentScheduler.WAITING_SET), anyString(), anyString()))
+        .thenReturn(Set.of("testAgentType"));
+
+    clusteredSortAgentScheduler.saturatePool();
+
+    // When parallelism is -1, the runningAgents is Optional.empty().
```
Comment (dbyron-sf) [inline cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:83]: But isn't parallelism initialized to 2 above?

File: cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java (hunk 6 above)
Comment (Young-Zen) [reply cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java:63]: Unnecessary.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java (hunk 5 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:117]: No need.

File: cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java (hunk 7 above)
Comment (Young-Zen) [reply cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java:83]: At first I thought that if parallelism is -1, both test cases will also pass. But think about it, these two test cases are used to test Semaphore's permit leaking, if parallelism is -1, there is no Semaphore.

**Review by dbyron-sf — APPROVED** [review]
